### PR TITLE
IOS-XR NetConf and Cliconf plugin work

### LINKED ISSUE #31345: [NETCONF]Create IOS-XR NetConf Plugin and module_utils refactor
##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
iosxr_*
module_utils/iosxr.py

##### ANSIBLE VERSION
```
2.5.0
```

##### SUMMARY
Add Netconf plugin for IOS-XR platform

### PR DESCRIPTION
**Pull request by kedarkekan**
- Netconf plugin addition for iosxr 
- Utilities refactoring to support netconf and cliconf
- iosx_banner refactoring for netconf and cliconf
- Integration testcases changes to accomodate above changes

##### SUMMARY
Fixes: #31345 #31278 #33324 (Task 1 of 6)

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
iosxr_*

##### ANSIBLE VERSION

```
2.5.0
```

### CODE REVIEW

File: lib/ansible/module_utils/iosxr.py [hunk 1]
```diff
@@ -26,12 +26,34 @@
...
 #
-from ansible.module_utils._text import to_text
-from ansible.module_utils.basic import env_fallback, return_values
+
+import json
+import sys
+from difflib import Differ
+from six import StringIO
+from io import BytesIO
+from lxml import etree
```
Comment (ganeshrn) [inline lib/ansible/module_utils/iosxr.py]: Please handle import error for third-party library import. Something similar to
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/junos.py#L31

File: lib/ansible/plugins/netconf/junos.py [hunk 2]
```diff
@@ -152,3 +152,39 @@ def reboot(self):
...
+
+    @ensure_connected
+    def get(self, *args, **kwargs):
+        try:
+            return self.m.get(*args, **kwargs).data_xml
+        except RPCError as exc:
+            raise Exception(to_xml(exc.xml))
+
+    @ensure_connected
+    def get_config(self, *args, **kwargs):
```
Comment (ganeshrn) [inline lib/ansible/plugins/netconf/junos.py:164]: Why is this changed required for junos plugin?

Comment (kedarkekan) [reply lib/ansible/plugins/netconf/junos.py:164]: abstracted out base class methods to handle platform specific behaviour.

File: lib/ansible/modules/network/iosxr/iosxr_banner.py [hunk 3]
```diff
@@ -155,20 +226,18 @@ def main():
...
 
-    commands = map_obj_to_commands((want, have), module)
-    result['commands'] = commands
-
-    if commands:
-        if not module.check_mode:
-            load_config(module, commands, result['warnings'], commit=True)
-        result['changed'] = True
+    if network_api == 'cliconf':
+        warnings.append('cli support for \'iosxr_banner\' is deprecated and will be removed 4 releases from 2.5. Use \'netconf\' instead')
```
Comment (ganeshrn) [inline lib/ansible/modules/network/iosxr/iosxr_banner.py]: You can use `module.deprecate` to throw a deprecation warning.

File: lib/ansible/modules/network/iosxr/iosxr_interface.py [hunk 4]
```diff
@@ -360,7 +360,7 @@ def main():
         enabled=dict(default=True, type='bool'),
         tx_rate=dict(),
         rx_rate=dict(),
-        delay=dict(default=10, type='int'),
+        delay=dict(default=5, type='int'),
```
Comment (ganeshrn) [inline lib/ansible/modules/network/iosxr/iosxr_interface.py]: Why is this changed to 5 sec? As per spec it should default 10 secs for all platforms

File: lib/ansible/module_utils/iosxr.py [hunk 5]
```diff
@@ -26,12 +26,43 @@
...
+from ansible.module_utils.netconf import NetconfConnection
+
+try:
+    from ncclient.xml_ import to_xml
+    HAS_NCCLIENT = True
+except ImportError:
+    HAS_NCCLIENT = False
+
+try:
+    from lxml import etree
```
Comment (privateip) [inline lib/ansible/module_utils/iosxr.py]: set import flag `HAS_LXML` for later checks to avoid exceptions begin raised

File: lib/ansible/module_utils/iosxr.py [hunk 6]
```diff
@@ -63,24 +98,270 @@ def check_args(module, warnings):
...
+        return reply
+    else:
+        print("utf8")
+        return reply.encode('UTF-8')
+
+
+# Note: Workaround for ncclient 0.5.3
+def remove_namespaces(rpc_reply):
+    xslt = transform_reply()
+    parser = etree.XMLParser(remove_blank_text=True)
```
Comment (privateip) [inline lib/ansible/module_utils/iosxr.py]: check if `HAS_LXML` is True

File: lib/ansible/module_utils/iosxr.py [hunk 7]
```diff
@@ -63,24 +98,270 @@ def check_args(module, warnings):
...
+# <config xmlns:xc="urn:ietf:params:xml:ns:netconf:base:1.0">
+#   <banners xmlns="http://cisco.com/ns/yang/Cisco-IOS-XR-infra-infra-cfg">
+#     <banner xc:operation="merge">
+#       <banner-name>motd</banner-name>
+#       <banner-text>Ansible banner example</banner-text>
+#     </banner>
+#   </banners>
+# </config>
+
+def build_xml_subtree(container_ele, xmap, param=None, opcode=None):
```
Comment (privateip) [inline lib/ansible/module_utils/iosxr.py]: check `HAS_LXML` flag

File: lib/ansible/module_utils/iosxr.py [hunk 8]
```diff
@@ -63,24 +98,270 @@ def check_args(module, warnings):
...
-    except KeyError:
-        rc, out, err = exec_command(module, cmd)
-        if rc != 0:
-            module.fail_json(msg='unable to retrieve current config', stderr=to_text(err, errors='surrogate_or_strict'))
-        cfg = to_text(out, errors='surrogate_or_strict').strip()
-        _DEVICE_CONFIGS[cmd] = cfg
+    container_ele.append(deepcopy(loop_root))
+
+
+def build_xml(container, xmap=None, params=None, opcode=None):
```
Comment (privateip) [inline lib/ansible/module_utils/iosxr.py]: check `HAS_LXML` flag

File: lib/ansible/module_utils/iosxr.py [hunk 9]
```diff
@@ -26,12 +26,43 @@
...
+import json
+import sys
+from difflib import Differ
+from io import BytesIO
+from copy import deepcopy
+
+from ansible.module_utils.six import StringIO
+from ansible.module_utils._text import to_text, to_bytes
+from ansible.module_utils.basic import env_fallback
 from ansible.module_utils.network_common import to_list, ComplexList
```
Comment (privateip) [inline lib/ansible/module_utils/iosxr.py]: remove `ComplexList`

File: lib/ansible/module_utils/iosxr.py [hunk 10]
```diff
@@ -63,24 +98,270 @@ def check_args(module, warnings):
...
+
+    return root
+
+
+def get_config_diff(module, running=None, candidate=None):
+    conn = get_connection(module)
+    capabilities = get_device_capabilities(module)
+    network_api = capabilities.get('network_api')
+
+    if network_api == 'cliconf':
```
Comment (privateip) [inline lib/ansible/module_utils/iosxr.py]: inconsistent returns from this function

**Review by privateip — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/iosxr.py [hunk 11]
```diff
@@ -40,10 +73,14 @@
     'password': dict(fallback=(env_fallback, ['ANSIBLE_NET_PASSWORD']), no_log=True),
     'ssh_keyfile': dict(fallback=(env_fallback, ['ANSIBLE_NET_SSH_KEYFILE']), type='path'),
     'timeout': dict(type='int'),
+    'authorize': dict(fallback=(env_fallback, ['ANSIBLE_NET_AUTHORIZE']), type='bool'),
```
Comment (gundalow) [inline lib/ansible/module_utils/iosxr.py]: Is this right, I didn't think iosxr had privilege mode.

Comment (kedarkekan) [reply lib/ansible/module_utils/iosxr.py]: right, it doesn't. will remove it.

File: lib/ansible/module_utils/iosxr.py (hunk 7 above)
Comment (kedarkekan) [reply lib/ansible/module_utils/iosxr.py]: fixed: added in parent method: build_xml()

File: test/integration/targets/iosxr_banner/tests/cli/basic-motd.yaml [hunk 12]
```diff
@@ -20,8 +22,8 @@
 - assert:
     that:
       - "result.changed == true"
-      - "'this is my motd banner' in result.commands[0]"
-      - "'that has a multiline' in result.commands[0]"
+      - "'this is my motd banner' in result.commands[1]"
```
Comment (gundalow) [inline test/integration/targets/iosxr_banner/tests/cli/basic-motd.yaml]: This might catch people out, maybe worth adding to `porting_guide_2.5.rst`

File: lib/ansible/modules/network/iosxr/iosxr_banner.py [hunk 13]
```diff
@@ -79,60 +81,140 @@
...
-    return {
-        'banner': module.params['banner'],
-        'text': text,
-        'state': module.params['state']
-    }
+from ansible.module_utils.network.iosxr.iosxr import get_device_capabilities, build_xml
+
+try:
+    from lxml import etree
+    HAS_LXML = True
```
Comment (gundalow) [inline lib/ansible/modules/network/iosxr/iosxr_banner.py]: Needs adding to `DOCUMENTATION`'s requirements.

File: lib/ansible/modules/network/iosxr/iosxr_banner.py [hunk 14]
```diff
@@ -155,20 +237,20 @@ def main():
...
 
-    commands = map_obj_to_commands((want, have), module)
-    result['commands'] = commands
-
-    if commands:
-        if not module.check_mode:
-            load_config(module, commands, result['warnings'], commit=True)
-        result['changed'] = True
+    if network_api == 'cliconf':
+        module.deprecate('cli support for \'iosxr_banner\' is deprecated. use netconf instead', version='4 releases from v2.5')
```
Comment (gundalow) [inline lib/ansible/modules/network/iosxr/iosxr_banner.py]: Use double quotes, then you will not need to escape
Also think about wording, perhaps:
`"cli support for 'iosxr_banner' is deprecated. Use transport 'netconf'  instead", version='4 releases from v2.5`

Is version an int or string?

File: lib/ansible/plugins/netconf/__init__.py [hunk 15]
```diff
@@ -162,7 +168,8 @@ def unlock(self, *args, **kwargs):
     def discard_changes(self, *args, **kwargs):
         """Revert the candidate configuration to the currently running configuration.
         Any uncommitted changes are discarded."""
-        return self.m.discard_changes(*args, **kwargs).data_xml
+        # return self.m.discard_changes(*args, **kwargs).data_xml
```
Comment (gundalow) [inline lib/ansible/plugins/netconf/__init__.py]: commented out code?

File: lib/ansible/plugins/netconf/__init__.py [hunk 16]
```diff
@@ -176,19 +183,30 @@ def commit(self, *args, **kwargs):
         :confirmed: whether this is a confirmed commit
         :timeout: specifies the confirm timeout in seconds
         """
-        try:
-            return self.m.commit(*args, **kwargs).data_xml
-        except RPCError as exc:
-            raise Exception(to_xml(exc.xml))
+        # try:
```
Comment (gundalow) [inline lib/ansible/plugins/netconf/__init__.py]: commented out code?

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/iosxr/iosxr_banner.py (hunk 14 above)
Comment (kedarkekan) [reply lib/ansible/modules/network/iosxr/iosxr_banner.py]: Version is a string. Will fix the wordings.

File: lib/ansible/plugins/netconf/__init__.py (hunk 15 above)
Comment (kedarkekan) [reply lib/ansible/plugins/netconf/__init__.py]: At multiple location in the base class methods, I commented out the code, while keeping it there as a reminder to restore it when all the platforms support `.data_xml` on ncclient and namespace removal. I can take them off and write a `TODO` somewhere if that looks good?

Comment (gundalow) [reply lib/ansible/plugins/netconf/__init__.py]: Maybe add
`#TODO Restore when all platforms support÷.data_xml`
Then have the commented out code?

Comment (kedarkekan) [reply lib/ansible/plugins/netconf/__init__.py]: Agree

File: test/integration/targets/iosxr_banner/tests/cli/basic-motd.yaml (hunk 12 above)
Comment (kedarkekan) [reply test/integration/targets/iosxr_banner/tests/cli/basic-motd.yaml]: ~~agree, will add a note.~~ Tweaking the code to change the behaviour to that of earlier.

File: lib/ansible/modules/network/iosxr/iosxr_banner.py [hunk 17]
```diff
@@ -16,32 +16,32 @@
...
     required: true
     default: null
     choices: ['login', 'motd']
   text:
     description:
-      - The banner text that should be
-        present in the remote device running configuration. This argument
-        accepts a multiline string, with no empty lines. Requires I(state=present).
+      - Banner text to be configured. Accepts multiline string, 
+      without empty lines. Requires I(state=present).
```
Comment (gundalow) [inline lib/ansible/modules/network/iosxr/iosxr_banner.py]: Indentation here doesn't look right

**Review by gundalow — APPROVED** [review]

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 18]
```diff
@@ -63,87 +102,282 @@ def check_args(module, warnings):
...
-        cmd = module.jsonify(cmd)
-        rc, out, err = exec_command(module, cmd)
-        if check_rc and rc != 0:
-            module.fail_json(msg=to_text(err, errors='surrogate_or_strict'), rc=rc)
-        responses.append(to_text(out, errors='surrogate_or_strict'))
-    return responses
+    return module.capabilities
+
+
+# Builds netconf xml rpc document from meta-data
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: Please add the comments as function doc

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 19]
```diff
@@ -63,87 +102,282 @@ def check_args(module, warnings):
...
+            for item in subtree_list:
+                container_ele.append(item)
+
+    return root
+
+
+def get_config_diff(module, running=None, candidate=None):
+    conn = get_connection(module)
+    capabilities = get_device_capabilities(module)
+    network_api = capabilities.get('network_api')
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: SInce network_api check is used in multiple lines can we move it out in another function?

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 20]
```diff
@@ -63,87 +102,282 @@ def check_args(module, warnings):
...
+    conn.discard_changes()
+
+
+def commit_config(module, comment=None, confirmed=False, confirm_timeout=None, persist=False, check=False):
+    conn = get_connection(module)
+    if check:
+        reply = conn.validate()
+    else:
+        capabilities = get_device_capabilities(module)
+        network_api = capabilities.get('network_api')
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: same as above

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 21]
```diff
@@ -63,87 +102,282 @@ def check_args(module, warnings):
...
+
+    return reply
+
+
+def get_config(module, source='running', config_filter=None):
+    global _DEVICE_CONFIGS
+
+    conn = get_connection(module)
+    capabilities = get_device_capabilities(module)
+    network_api = capabilities.get('network_api')
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: same as above

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 22]
```diff
@@ -63,87 +102,282 @@ def check_args(module, warnings):
...
+
+        cfg = to_text(out, errors='surrogate_then_replace').strip()
+        _DEVICE_CONFIGS.update({key: cfg})
+        return cfg
+
+
+def load_config(module, command_filter, warnings, replace=False, admin=False, commit=False, comment=None):
+    conn = get_connection(module)
+    capabilities = get_device_capabilities(module)
+    network_api = capabilities.get('network_api')
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: same as above

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 23]
```diff
@@ -63,87 +102,282 @@ def check_args(module, warnings):
...
+
+    if network_api == 'netconf':
+        if not HAS_NCCLIENT:
+            module.fail_json(msg=('ncclient is not installed'))
+        if not HAS_XML:
+            module.fail_json(msg=('lxml is not installed'))
+
+        # ret = conn.lock(target = 'candidate')
+        # ret = conn.discard_changes()
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: Is the commented out code required?

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 24]
```diff
@@ -63,87 +102,282 @@ def check_args(module, warnings):
...
+        try:
+            out = conn.edit_config(etree.tostring(command_filter))
+        finally:
+            # ret = conn.unlock(target = 'candidate')
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: Same here

File: lib/ansible/modules/network/iosxr/iosxr_banner.py [hunk 25]
```diff
@@ -79,60 +79,140 @@
...
+        self._banners_meta = collections.OrderedDict()
+        self._banners_meta.update([
+            ('banner', {'xpath': 'banners/banner', 'tag': True, 'attrib': "operation"}),
+            ('a:banner', {'xpath': 'banner/banner-name'}),
+            ('a:text', {'xpath': 'banner/banner-text', 'operation': 'edit'})
+        ])
+
+    def map_obj_to_commands(self):
+        state = self._module.params['state']
+        _get_filter = build_xml('banners', xmap=self._banners_meta, params=self._module.params, opcode="filter")
```
Comment (ganeshrn) [inline lib/ansible/modules/network/iosxr/iosxr_banner.py:166]: I think it is better to pass module as argument to build_xml() and handle lxml install check-in module_utils instead of having it in each module

File: lib/ansible/modules/network/iosxr/iosxr_banner.py [hunk 26]
```diff
@@ -79,60 +79,140 @@
...
+        if state == 'absent' and banner_name == self._module.params['banner'] and len(banner_text):
+            opcode = "delete"
+        elif state == 'present':
+            opcode = 'merge'
+
+        self._result['commands'] = []
+        if opcode:
+            _edit_filter = build_xml('banners', xmap=self._banners_meta, params=self._module.params, opcode=opcode)
+            if _edit_filter == 'no_lxml':
+                self._module.fail_json(msg='lxml is not installed')
```
Comment (ganeshrn) [inline lib/ansible/modules/network/iosxr/iosxr_banner.py]: Same as above

File: lib/ansible/modules/network/iosxr/iosxr_command.py [hunk 27]
```diff
@@ -125,9 +126,9 @@
 import time
 
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.network.iosxr.iosxr import run_commands, iosxr_argument_spec, check_args
+from ansible.module_utils.network.iosxr.iosxr import run_command, iosxr_argument_spec
+from ansible.module_utils.network.iosxr.iosxr import check_args, command_spec
```
Comment (ganeshrn) [inline lib/ansible/modules/network/iosxr/iosxr_command.py]: check_arg() call can be removed from all iosx module and utils and it is not longer used

File: lib/ansible/modules/network/iosxr/iosxr_command.py [hunk 28]
```diff
@@ -174,6 +175,8 @@ def main():
 
     spec.update(iosxr_argument_spec)
 
+    spec.update(command_spec)
```
Comment (ganeshrn) [inline lib/ansible/modules/network/iosxr/iosxr_command.py:178]: Adding `command_spec`  as part of top-level spec is not the right way to use of `ComplexList

File: lib/ansible/modules/network/iosxr/iosxr_banner.py (hunk 25 above)
Comment (kedarkekan) [reply lib/ansible/modules/network/iosxr/iosxr_banner.py:166]: Module object is of no use in build_xml. instead moved out lib install check in module.

File: lib/ansible/plugins/action/iosxr.py [hunk 29]
```diff
@@ -48,7 +48,17 @@ def run(self, tmp=None, task_vars=None):
         elif self._play_context.connection == 'local':
             provider = load_provider(iosxr_provider_spec, self._task.args)
             pc = copy.deepcopy(self._play_context)
-            pc.connection = 'network_cli'
+            if self._task.action in ['iosxr_netconf', 'iosxr_config', 'iosxr_command'] or \
```
Comment (ganeshrn) [inline lib/ansible/plugins/action/iosxr.py:51]: iosxr_config and iosxr_command should support both netconf and cli?

File: lib/ansible/plugins/connection/netconf.py [hunk 30]
```diff
@@ -114,6 +114,7 @@ def __init__(self, play_context, new_stdin, *args, **kwargs):
         self._netconf = None
         self._manager = None
         self._connected = False
+        self._netconf = None
```
Comment (ganeshrn) [inline lib/ansible/plugins/connection/netconf.py]: This line is not required as _netconf is set to None in L114

File: lib/ansible/modules/network/iosxr/iosxr_command.py (hunk 28 above)
Comment (kedarkekan) [reply lib/ansible/modules/network/iosxr/iosxr_command.py:178]: As discussed, will keep it as is; for later modification in #33575

**Review by ganeshrn — COMMENTED** [review]
Please add netconf related integration test for modules that support netconf transport

File: lib/ansible/plugins/action/iosxr.py (hunk 29 above)
Comment (kedarkekan) [reply lib/ansible/plugins/action/iosxr.py:51]: iosxr does not support cli commands over netconf. Keeping support of xml rpc schema as input at a later priority over individual config modules.

File: lib/ansible/module_utils/network/iosxr/iosxr.py (hunk 23 above)
Comment (kedarkekan) [reply lib/ansible/module_utils/network/iosxr/iosxr.py]: Required, but the lock isn't stable on the box, hence commented for a later todo.

File: lib/ansible/module_utils/network/iosxr/iosxr.py (hunk 19 above)
Comment (kedarkekan) [reply lib/ansible/module_utils/network/iosxr/iosxr.py]: This may move out a couple of lines only.

File: lib/ansible/module_utils/network/iosxr/iosxr.py (hunk 23 above)
Comment (ganeshrn) [reply lib/ansible/module_utils/network/iosxr/iosxr.py]: In that case please add a TODO in comment.

File: lib/ansible/modules/network/iosxr/iosxr_banner.py (hunk 25 above)
Comment (ganeshrn) [reply lib/ansible/modules/network/iosxr/iosxr_banner.py:166]: Having this check in each module will result in code duplication and an overhead for module writer. `module` param will be used if the check is moved to utils function

Comment (kedarkekan) [reply lib/ansible/modules/network/iosxr/iosxr_banner.py:166]: build_xml may not have access to the `module` always (like from plugins); so moving the install check to utils for modules and plugins will do a separate install check.

**Review by privateip — APPROVED** [review]

**Review by ganeshrn — APPROVED** [review]
